spec/features/two_factor_authentication/change_factor_spec.rb

```diff
@@ -53,9 +53,6 @@
       expect(
         user.reload.phone_configuration.confirmed_at
       ).to_not eq(@previous_phone_confirmed_at)
-
-      visit login_two_factor_path(otp_delivery_preference: 'sms')
```

Comment:
I'm on my phone at the moment, but does a `git blame` point to the PR these lines were added in?

Reply:
@monfresh could you elaborate on this PR?  The code has changed dramatically in 2 years so I'm not sure if the issue you were seeing still applies or if it is a valid scenario a user would encounter while attempting to change phone. Your input would be greatly appreciated. Thanks!  https://github.com/18F/identity-idp/commit/68be2af0a816ba7135bede318349d579ef7787d9#diff-5bc12837c9dd82673bc7fd51b5d0aec7

Reply:
Yes, here's the PR that added this test: https://github.com/18F/identity-idp/pull/647
At the time, the bug was that if you tried accessing the OTP verification page after confirming a phone, you were able to. By adding these tests, it made sure that fully authenticated users who had just confirmed a phone were redirected instead of being allowed to view the OTP verification page.

So, this test should remain, and I would also add a new one for the scenario where a user comes from an SP (both SAML and OIDC).

Reply:
If you look at the spec there is already a test to make sure they get redirected back to the account page after changing the phone.  How would they then get to the otp verification page from the account page?

Reply:
I see that this test is failing with the change in this PR. This is because the logic now goes through `after_sign_in_path_for(_user)`, which first checks `user_session[:stored_location]`, and in the scenario where the user waits long enough before changing their number such that they are prompted to enter their password + 2FA, we set the stored_location to the  "manage phone" page so that they are redirected there after confirming their password and 2FA. But since we never clear the stored location after they're done, it says in the session, and if they try to access a page that ends up going through `after_sign_in_path_for(_user)`, they will end up on the manage phone page instead of their account page.

Reply:
Yes.  I can definitely fix it but I was debating whether I should just to satisfy a test that can never occur.  My question is really do we want to fix them all if they are not paths that a user can actually visit (ie valid scenarios)?  If so I can surely turn up some more like this one when traversing all the states of the conditionals when used in the context of the ui.  We have at least a half dozen of those conditions now in prod that regularly get triggered only by pen testers (500's).  At least this one just takes you back to the manage phone.

Reply:
The second tab scenario occurs in prod when the user initiates login from the sp not when we do account initiated actions.  However, it took me longer to type all this then actually put out fix to satisfy this arbitrary test hence I'll just fix it :-)

Reply:
Consider this scenario:

1. In one tab, visit the sign in page directly and sign in with email and password, leaving the 2FA confirmation page up but without submitting it
2. In a second tab, do the same thing, but this time, submit the 2FA form
3. Wait a couple of minutes
4. Click on "Edit" next to your phone number on your account page
5. You should be prompted to enter your password, enter it
6. You should be prompted to 2FA on your existing phone, submit the 2FA form
7. You should now be on the manage phone page, where you can either enter a new number, or just submit the form with your existing number. It doesn't matter for the purposes of this test.
8. You should now be on your account page as expected
9. Now go to the other tab that still has the 2FA form displayed and submit the form
Expect Result: you are on your account page
Actual Result: you are on the manage phone page

The fix is simple: change `user_session[:stored_location]` to `user_session.delete(:stored_location)`

Reply:
And how will this work for the third tab after you delete the stored location?  The argument has infinite regress...

Reply:
That's just one example, but it doesn't have to be that involved. Any page that a user can visit directly is a valid test. You literally have the exact same test in the new test you created for this PR:
```ruby
visit login_two_factor_path(otp_delivery_preference: 'sms')
```
In your case, you're checking that if the user came from an SP and fully authenticated, if they try to visit the 2FA confirmation page directly, they cannot view that page and are instead redirected to the appropriate place based on the current context.

The test in `change_factor_spec` is the same test, but in a different context.

Reply:
I'm not sure what you mean about the third tab. Can you provide steps to reproduce?

Reply:
That stored_location is used a lot.  I'm also concerned about the side effects it will produce if we delete it after first use.  It apparently is deleted nowhere else in the code.  Bizarre.

Reply:
I wasn't saying your test was invalid I was saying the scenario is not likely to occur.  I'm using the exact same test for a scenario that is not only likely to happen but is happening daily to our users.  Again the key distinction is logging in from the sp.  This creates new windows and/or tabs.  The scenario you are describing would have to be initiated on login.gov and from the account page and require them to switch back and forth between tabs.  The solution as presented here fixes all scenarios not just for the otp screen, but the options screen and any other 2fa screen after login from the sp but before authentication.  Users were getting routed to the account page in every scenario on gets as well as posts.  I will fix the scenario you describe.  However, the code is definitely precarious and this one before_filter was used as a catch all for a dozen different scenarios.  Once again we'll have to rely on the completeness of our feature specs which have burned us a few times.

Reply:
Actually it's not the contexts of the before_filter it's the conditionals in the before_filter.  They should be broken out into separate filters.

Reply:
The stored location is not used a lot. It is literally only used for the scenario where a user is trying to change their phone, password, or email, and it's been long enough since they've last authenticated that they are required to confirm their current password and 2FA first before being able to make those changes.

Perhaps you're thinking of Devise's `stored_location_for`? We're not using that Devise feature.

The point of access guards is to prevent the user from seeing a page that doesn't or no longer applies to them. For example, a user who is already verified at LOA3 should not be able to go through verification again by visiting the `/verify` endpoint directly, which is why we redirect to `/verify/activated`. Is it likely that a user will try to visit that endpoint directly? Perhaps not. Is it possible? It sure is, and we must prevent it and make sure they get redirected to the appropriate place.

Reply:
Yep.  We could even mechanize the testing and just spin through all the routes.  However, my opportunity costs are getting deeper and deeper on this one and I'm getting diminishing returns...